### PR TITLE
Parse YAPF diffs into TextEdits (instead of sending the full doc)

### PR DESCRIPTION
**Pull request by masad-frost**
Currently formatting with YAPF sends back the full document as a range, it's not great because:
- Editors that allow edits while formatting is pending cannot reasonably consolidate the format diff with the user's local changes
- Editors that have to talk to servers now have to send the whole document back and forth (admittedly, for smaller documents the added JSON structures offset that)
- The person who built the plugin wanted to do it, but couldn't be bothered :laughing: 
- In general, it is the recommended approach by the spec

To support this feature I used YAPF's unified diff output option. I took the output and passed it to [`whatthepatch`](https://github.com/cscorley/whatthepatch) to produce a parsed result, the format is outlined in the library's readme. We iterate over lines in the diff and create `TextEdit`s. Left some comments around the code to make it more readable.

To be able to test without strictly matching `TextEdit`s I also added a function to the `workspace` module called `apply_text_edits` which takes a `Document` instance and `TextEdit`s and returns a new source/string with the text edits applied. I wrote tests for this function based on the tests in the vscode language server repo. This function is now only used for testing, but I imagine it being useful for anyone working with text edits

### CODE REVIEW

**Comment by krassowski** [comment]
Thanks for starting this work!

The tests are failing on CI - it looks like they need updating to match the changes in 341732d:

```
>       assert doc.apply_text_edits(res) == "A = ['h', 'w', 'a']\n\nB = ['h', 'w']\n"
E       AttributeError: 'Document' object has no attribute 'apply_text_edits'
```

**Comment by masad-frost** [comment]
@krassowski the tests pass locally now if you wanna run the workflow.

**Comment by masad-frost** [comment]
Actually fixed now. Ran `pytest` and `pylint`

**Comment by masad-frost** [comment]
Weird, I'm not getting these lint issues locally. Will look into it

**Comment by masad-frost** [comment]
Will merge main after https://github.com/python-lsp/python-lsp-server/pull/134 is merged

File: pylsp/workspace.py [hunk 1]
```diff
@@ -129,6 +129,84 @@ def _create_document(self, doc_uri, source=None, version=None):
         )
 
 
+def get_well_formatted_range(range):
```
Comment (ccordoba12) [inline pylsp/workspace.py]: All functions you added here are not used directly in this file, but only in our tests. So I don't understand why they are here and their usage.

Is not `apply_text_edits` already implemented in the `apply_change` method of the `Document` class?

**Comment by ccordoba12** [comment]
@masad-frost, now that #134 is merged, you can fix the merge conflicts here so we can merge it.

File: pylsp/workspace.py (hunk 1 above)
Comment (masad-frost) [reply pylsp/workspace.py]: `apply_change` handles [TextDocumentContentChangeEvent](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentContentChangeEvent) where as `apply_text_edits` handles [TextEdit](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textEdit). They have different data structure and semantics in the protocol.

Happy to move them to tests, but thought it might be useful to have a function that handles `TextEdit`s. Might be a YAGNI, happy to move to test file.

Comment (ccordoba12) [reply pylsp/workspace.py]: > apply_change handles TextDocumentContentChangeEvent where as apply_text_edits handles TextEdit

Ok, I see. Then please move all these functions to `pylsp/text_edits.py`.

**Review by ccordoba12 — COMMENTED** [review]
Hey @masad-frost, thanks for patience with this! Last comment, then this should be ready.

Also, please fix the merge conflicts that this PR has now.

File: pylsp/workspace.py (hunk 1 above)
Comment (masad-frost) [reply pylsp/workspace.py]: You got it

File: pylsp/_utils.py [hunk 2]
```diff
@@ -228,8 +228,8 @@ def is_process_alive(pid):
 
 
 def get_eol_chars(text):
-    """Get EOL chars used in text."""
+    """Get EOL chars used in text. Defaults to line feed"""
     match = EOL_REGEX.search(text)
     if match:
         return match.group(0)
-    return None
+    return "\n"
```
Comment (masad-frost) [inline pylsp/_utils.py]: I checked usage sites, this seems to make sense

File: pylsp/plugins/yapf_format.py [hunk 3]
```diff
@@ -36,75 +38,162 @@ def pylsp_format_range(document, range, options=None):  # pylint: disable=redefi
...
+                    }
+                },
+                'newText': ''
+            })
+            prev_line_no = lsp_line_no
+
+    return text_edits
+
+
+def ensure_eof_new_line(document, eol_chars, text_edits):
```
Comment (masad-frost) [inline pylsp/plugins/yapf_format.py:142]: This is a bit of a hack, see: https://github.com/google/yapf/issues/1008

**Comment by masad-frost** [comment]
@ccordoba12 lint issues seem to be unrelated from this branch.

image

File: pylsp/_utils.py [hunk 4]
```diff
@@ -14,7 +14,7 @@
 JEDI_VERSION = jedi.__version__
 
 # Eol chars accepted by the LSP protocol
-EOL_CHARS = ['\r\n', '\r', '\n']
+EOL_CHARS = ['\n', '\r\n', '\r']
```
Comment (ccordoba12) [inline pylsp/_utils.py]: The ordering here was done for performance reasons.So, please revert this change.

File: pylsp/_utils.py (hunk 2 above)
Comment (ccordoba12) [reply pylsp/_utils.py]: I'm sorry but this function is used by two other projects besides this one, so you can't modify it like this now.

File: pylsp/text_edit.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
+def get_well_formatted_range(lsp_range):
```
Comment (ccordoba12) [inline pylsp/text_edit.py:4]:
```suggestion
# Copyright 2017-2020 Palantir Technologies, Inc.
# Copyright 2021- Python Language Server Contributors.

def get_well_formatted_range(lsp_range):
```

Missing header

File: test/test_text_edit.py [hunk 6]
```diff
@@ -0,0 +1,342 @@
+from pylsp.text_edit import OverLappingTextEditException, apply_text_edits
```
Comment (ccordoba12) [inline test/test_text_edit.py:4]:
```suggestion
# Copyright 2017-2020 Palantir Technologies, Inc.
# Copyright 2021- Python Language Server Contributors.

from pylsp.text_edit import OverLappingTextEditException, apply_text_edits
```

Missing header.

File: test/test_workspace.py [hunk 7]
```diff
@@ -5,7 +5,6 @@
 import pytest
 from pylsp import uris
 
-
```
Comment (ccordoba12) [inline test/test_workspace.py]: Please revert this change because it's not really necessary.

**Review by ccordoba12 — COMMENTED** [review]
Hey @masad-frost, thanks for the updates! I think you're really close to finish this one.

**Comment by ccordoba12** [comment]
> @ccordoba12 lint issues seem to be unrelated from this branch.

I'll help with that after you solve my comments.

File: pylsp/_utils.py (hunk 4 above)
Comment (masad-frost) [reply pylsp/_utils.py]: Seems like it's worth a comment in that case

File: pylsp/_utils.py (hunk 2 above)
Comment (masad-frost) [reply pylsp/_utils.py]: Oh I wasn't aware that this project had contracts beyond the wire protocol.

File: pylsp/text_edit.py (hunk 5 above)
Comment (masad-frost) [reply pylsp/text_edit.py:4]: Will fix it up.

However, you should consider removing the copyright headers, unless we're distributing files individually, they add noise to MIT projects for no return. The License.md file should be good enough unless a specific file has different licensing from the root license.

**Comment by masad-frost** [comment]
thanks, quick fixes esp with the suggestions!

File: pylsp/_utils.py (hunk 4 above)
Comment (ccordoba12) [reply pylsp/_utils.py]: Agreed. I saw that you added it, so thanks for that.

File: pylsp/text_edit.py (hunk 5 above)
Comment (ccordoba12) [reply pylsp/text_edit.py:4]: You're probably right, but I think it's easier to add this header for new files than remove them from all of them.

**Comment by masad-frost** [comment]
Just the lints are left! Let me know if there's anything else I can address. If you have more test cases in mind or something

**Review by ccordoba12 — APPROVED** [review]
Thanks for your hard on this @masad-frost! I think it's finally ready.
